### PR TITLE
MainWindow refactor and resizable sections list

### PR DESCRIPTION
**Pull request by drognanar**

### CODE REVIEW

File: src/core/settings.cpp [hunk 1]
```diff
@@ -117,9 +117,17 @@ void Settings::load()
     }
     m_settings->endGroup();
 
+
```
Comment (trollixx) [inline src/core/settings.cpp]: whitespace

File: src/core/settings.cpp [hunk 2]
```diff
@@ -117,9 +117,17 @@ void Settings::load()
...
+
     m_settings->beginGroup(GroupState);
     windowGeometry = m_settings->value(QStringLiteral("window_geometry")).toByteArray();
-    splitterGeometry = m_settings->value(QStringLiteral("splitter_geometry")).toByteArray();
+    verticalSplitterGeometry = m_settings->value(QStringLiteral("splitter_geometry")).toByteArray();
+
+    sectionsSplitterSizes = QList<int>();
+    QList<QVariant> splitterSizes = m_settings->value(QStringLiteral("sections_geometry")).toList();
+    for (int i = 0; i < splitterSizes.size(); i++)
```
Comment (trollixx) [inline src/core/settings.cpp]: Please use range-based for here

File: src/core/settings.cpp [hunk 3]
```diff
@@ -164,9 +172,14 @@ void Settings::save()
     m_settings->endGroup();
 #endif
 
+    QList<QVariant> splitterSizes;
+    for (int i = 0; i < sectionsSplitterSizes.size(); i++)
```
Comment (trollixx) [inline src/core/settings.cpp]: Same as above

File: src/ui/mainwindow.cpp [hunk 4]
```diff
@@ -104,9 +104,13 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
             ui->lineEdit, static_cast<void (SearchEdit::*)()>(&SearchEdit::setFocus));
 
     restoreGeometry(m_settings->windowGeometry);
-    ui->splitter->restoreState(m_settings->splitterGeometry);
+    ui->splitter->restoreState(m_settings->verticalSplitterGeometry);
     connect(ui->splitter, &QSplitter::splitterMoved, [=](int, int) {
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: Hm, actually we do store the geometry in `closeEvent()`, maybe it's excessive to update it on every move.

File: src/ui/mainwindow.cpp [hunk 5]
```diff
@@ -104,9 +104,13 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
...
-        m_settings->splitterGeometry = ui->splitter->saveState();
+        m_settings->verticalSplitterGeometry = ui->splitter->saveState();
+    });
+    connect(ui->sectionsSplitter, &QSplitter::splitterMoved, [=](int, int) {
+        if (ui->sections->isVisible())
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: How about storing the height of `sections` in `displaySections()` before hiding it?

File: src/ui/mainwindow.cpp [hunk 6]
```diff
@@ -468,6 +461,38 @@ void MainWindow::createTab()
...
+
+    if (!searchState->searchQuery.isEmpty()) {
+        ui->treeView->setModel(searchState->zealSearch);
+        ui->treeView->setRootIsDecorated(false);
+        ui->treeView->reset();
+    } else {
+        ui->treeView->setModel(m_zealListModel);
+        ui->treeView->setColumnHidden(1, true);
+        ui->treeView->setRootIsDecorated(true);
+        ui->treeView->reset();
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: Duplicated in the above block, so should be moved out.

File: src/ui/mainwindow.h [hunk 7]
```diff
@@ -122,9 +122,12 @@ private slots:
 
 private:
     void displayViewActions();
+    void displayTreeView();
+    void displaySections();
     void setupSearchBoxCompletions();
     void reloadTabState();
     void displayTabs();
+    SearchState *currentSearchState();
```
Comment (trollixx) [inline src/ui/mainwindow.h]: `const` is missing.

**Comment by trollixx** [comment]
Mostly nitpicking, but I've spent so much time cleaning up the code base, so now I prefer to keep it pretty in the first place :)

In the commit description it should be either `MainWindow` or `main window`. And in the description either `Move logic...` or `This change moves...`. Commit comments should be in the imperative form.

Actually, the name `sections` does not carry much sense. It should be in fact `toc` and work differently (see #386). I have a renaming change, but now I'll wait until this change lands.

**Comment by drognanar** [comment]
Updated PR. I'll rename the commits once everything else is fine.

Regarding the name for `sections` I am unsure about calling it `toc` either as the `treeView` acts more like a table of contents.

Also I am well aware that it should work differently. I made more progress on my master branch and one of commits adds support for .dashtoc files 969dfb4d9c8. However, I'm waiting for the current PRs to complete first.

**Comment by trollixx** [comment]
Thanks for being patient. The resulting code looks good to me. Can you amend the relevant commits instead of pushing fixes on top of them?

Perhaps `pageToc` would work, but that's not important at the moment. Unfortunately, we don't have any real table of contents in docsets (#447).

**Comment by drognanar** [comment]
I rebased the change. I normally do that after the fixes get accepted to make the review easier.

**Comment by trollixx** [comment]
Yeah, it all looks good. Any idea why it has picked up my latest commit?

**Comment by trollixx** [comment]
`git rebase master` fixes the problem in my local branch.

**Comment by drognanar** [comment]
I rebased against upstream before and somehow it did not pick up one of the commits.
Fixed.

**Pull request merged**